lib/wab/impl/handler.rb

```diff
@@ -15,39 +15,48 @@ def initialize(server, shell)
       end
 
       def do_GET(req, res)
-        ctrl, path, query = extract_req(req)
-        log_response('controller.read', path, query) if @shell.logger.info?
-        send_result(compute_result('GET', ctrl, path, query), res)
-      rescue Exception => e
-        send_error(e, res)
+        handle('GET', req, res)
       end
 
       def do_PUT(req, res)
-        ctrl, path, query, body = extract_req(req)
-        log_response('controller.create', path, query, body) if @shell.logger.info?
-        send_result(compute_result('PUT', ctrl, path, query, body), res)
-      rescue Exception => e
-        send_error(e, res)
+        handle('PUT', req, res)
       end
 
       def do_POST(req, res)
-        ctrl, path, query, body = extract_req(req)
-        log_response('controller.update', path, query, body) if @shell.logger.info?
-        send_result(compute_result('POST', ctrl, path, query, body), res)
-      rescue Exception => e
-        send_error(e, res)
+        handle('POST', req, res)
       end
 
       def do_DELETE(req, res)
-        ctrl, path, query = extract_req(req)
-        log_response('controller.delete', path, query) if @shell.logger.info?
-        send_result(compute_result('DELETE', ctrl, path, query), res)
-      rescue Exception => e
-        send_error(e, res)
+        handle('DELETE', req, res)
       end
 
       private
 
+      # core function that handles the commonly-used HTTP methods:
```

Comment:
This code still uses if and elsif  to look up the code to run. That is horribly inefficient. I see that you want to combine the extract_req, the log, and the rescue into one reusable function but that is not the way to do it. If you count the lines it is about the same between both options anyway. Repeating a code pattern is sometimes the most easily understood and best performing. The suggest changes do not make the code easier to follow or any more dry.

A more efficient approach would be to pass a block to the common function but even that is much less efficient than just calling the function like controller.read directly.